### ISSUE TITLE
Missing Instantiations

### ISSUE THREAD

**Issue by sairajat** [dealii/dealii#5246]
Hello,
I am using dealii to solve a 3D solid mechanics problem.  The code works fine in 3D and now I am trying to make it work for 2D as well. I am doing this by using dim = 2 and spacedim = 3 for all deal.ii classes.
I am doing this because I want the Points and fe_values gradients 

However, on compiling the code, I find several undefined references for the following three functions.
```
error: undefined reference to 'void dealii::FETools::compute_projection_from_quadrature_points_matrix<2, 3>(dealii::FiniteElement<2, 3> const&, dealii::Quadrature<2> const&, dealii::Quadrature<2> const&, dealii::FullMatrix<double>&)'

error: undefined reference to 'void dealii::VectorTools::compute_no_normal_flux_constraints<2, 3, dealii::DoFHandler>(dealii::DoFHandler<2, 3> const&, unsigned int, std::set<unsigned char, std::less<unsigned char>, std::allocator<unsigned char> > const&, dealii::ConstraintMatrix&, dealii::Mapping<2, 3> const&)'

error: undefined reference to 'void dealii::DoFTools::make_zero_boundary_constraints<2, 3, dealii::DoFHandler>(dealii::DoFHandler<2, 3> const&, unsigned char, dealii::ConstraintMatrix&, dealii::ComponentMask const&)'

```
So, I looked into the instantiation file for each of them and found that these instantiations are missing while the documentation says that these templates can be used.

If I am correct, this can be solved by instantiating these functions in the corresponding files.
Please let me know. Thanks.

Links to file:

[Function 1](https://github.com/dealii/dealii/blob/4905a5047535b65d9ec5b625c3efeb5b4936e9f5/source/fe/fe_tools.inst.in#L193)

[Function 2 ](https://github.com/dealii/dealii/blob/5a9946462bbb94d1f4426bd8d42e33466eb78b25/source/numerics/vector_tools_constraints.inst.in#L62)

[Function 3](https://github.com/dealii/dealii/blob/d58dd0bbcb559af07b01cd3406cb6f6aa56fc57b/source/dofs/dof_tools_constraints.inst.in#L86)

**Comment by drwells** [comment]
Thanks for reporting these: your diagnosis is exactly right.

**Comment by sairajat** [comment]
@drwells  Thanks for replying.
May I edit the files and create a PR?

**Comment by drwells** [comment]
I would be grateful if you would be willing to spend the time to fix the first two (I went ahead and tackled the third one; see #5247). Let me know if you have any questions; I would be happy to help (as would the other developers).

This might not be trivial; I suspect that the original author of `vector_tools_constraints.inst.in` had a good reason to skip the nonzero codimension cases at the time he or she wrote line 23, but our support for these things has gotten better over the years and things *should* work.

**Comment by drwells** [comment]
@sairajat If you don't have time to work on this then one of us can handle it. Would it be possible for you to still contribute a test based on your code?

**Comment by sairajat** [comment]
@drwells  Yes, I dont think I currently can take out any time to work on this bug.
I will still see if I can come up with a MWE to contribute a test.

**Comment by masterleinad** [comment]
#5476 fixes the first case so that we are left with `VectorTools::compute_no_normal_flux_constraints` for which I don't see a straightforward solution.

**Comment by drwells** [comment]
I fixed about half of the `dim` vs `spacedim` conflicts. I can finish this one up.

**Comment by masterleinad** [comment]
For me it's not quite clear how to implement these normal flux constraints for `dim!=spacedim` but I look forward to your solution. 
Maybe putting an appropriate `ExcNotImplemented` would also be fine for the moment.

**Comment by drwells** [comment]
I haven't spent long thinking about it (I wanted to get things compiling first before dealing with the hard stuff) but you're probably right; I suspect these constraints don't make a lot of sense in the nonzero codimension case.

**Comment by bangerth** [comment]
I think they still do -- it's just a question of how you define the "normal vector" to the boundary in the case of a codim-1 situation. To me that's a vector that's perpendicular to the boundary and tangential to the manifold you're working on.  The more interesting question may be whether you apply this to a vector field that's `dim` dimensional or `spacedim` dimensional.

**Comment by sairajat** [comment]
@drwells  Just a minor update. I will be able to provide a MWE by the end of this weekend.
Thanks for the patience.

**Comment by drwells** [comment]
I haven't had time to work on the missing instantiations. The test will be quite useful though :)

**Comment by sairajat** [comment]
[BugFinder.zip](https://github.com/dealii/dealii/files/1507055/BugFinder.zip)

Please find attached the code that gives error on linking when run with dim = 2 and spacedim = 3.
It runs fine when dim = spacedim = 3.

Also, please note, for the test case, I have set dim = 3 (which is like spacedim). 
Mesh_dim is the variable that changes the dimension in my case. So, it runs when meshdim = 3,
but gives linking error when meshdim = 2.

Thanks.
